src/qibolab/instruments/qm/program/sweepers.py

```diff
@@ -0,0 +1,187 @@
+import math
+from dataclasses import dataclass
+from typing import Optional
+
+import numpy as np
+import numpy.typing as npt
+from qibo.config import raise_error
+from qm import qua
+from qm.qua import declare, fixed
+from qm.qua._dsl import _Variable  # for type declaration only
+
+from qibolab.components import Config
+from qibolab.sweeper import Parameter, Sweeper
+
+from ..config import operation
+from .arguments import ExecutionArguments
+
+MAX_OFFSET = 0.5
+"""Maximum voltage supported by Quantum Machines OPX+ instrument in volts."""
+
+
+def maximum_sweep_value(values: npt.NDArray, value0: npt.NDArray) -> float:
+    """Calculates maximum value that is reached during a sweep.
+
+    Useful to check whether a sweep exceeds the range of allowed values.
+    Note that both the array of values we sweep and the center value can
+    be negative, so we need to make sure that the maximum absolute value
+    is within range.
+
+    Args:
+        values (np.ndarray): Array of values we will sweep over.
+        value0 (float, int): Center value of the sweep.
+    """
+    return max(abs(min(values) + value0), abs(max(values) + value0))
+
+
+def check_max_offset(offset: Optional[float], max_offset: float = MAX_OFFSET):
+    """Checks if a given offset value exceeds the maximum supported offset.
+
+    This is to avoid sending high currents that could damage lab
+    equipment such as amplifiers.
+    """
+    if max_offset is not None and abs(offset) > max_offset:
+        raise_error(
+            ValueError, f"{offset} exceeds the maximum allowed offset {max_offset}."
+        )
+
+
+# def _update_baked_pulses(sweeper, qmsequence, config):
+#    """Updates baked pulse if duration sweeper is used."""
+#    qmpulse = qmsequence.pulse_to_qmpulse[sweeper.pulses[0].id]
+#    is_baked = isinstance(qmpulse, BakedPulse)
+#    for pulse in sweeper.pulses:
+#        qmpulse = qmsequence.pulse_to_qmpulse[pulse.id]
+#        if isinstance(qmpulse, BakedPulse):
+#            if not is_baked:
+#                raise_error(
+#                    TypeError,
+#                    "Duration sweeper cannot contain both baked and not baked pulses.",
+#                )
+#            values = np.array(sweeper.values).astype(int)
+#            qmpulse.bake(config, values)
+
+
+@dataclass
+class QuaSweep:
+    """Base class for different sweep parameters."""
+
+    sweeper: Sweeper
+
+    def declare(self) -> _Variable:
+        """Declares the QUA variable that is swept."""
+        return declare(fixed)
```

Comment:
For this, you just need a mapping: `Sweeper.parameter -> qua_variable_type`.

Reply:
f6d3b5c16d366e2bef082d90a9ea4956a255b5ab